python/ray/train/callbacks/results_preprocessors/aggregate.py

```diff
@@ -0,0 +1,108 @@
+import logging
+from typing import Dict, List, Optional
+
+import numpy as np
+
+from ray.train.callbacks.results_preprocessors.preprocessor import ResultsPreprocessor
+from ray.train.callbacks.results_preprocessors.aggregate_fn import (
+    AggregateFn,
+    Average,
+    Max,
+    WeightedAverage,
+    VALID_AGGREGATE_TYPES,
+)
+
+logger = logging.getLogger(__name__)
+
+
+def _get_values(key, reported_metrics, results):
+    """Return values in the results list from all workers."""
+    if key not in reported_metrics:
+        logger.warning(
+            f"`{key}` is not reported from workers, so it is ignored. "
+            "Please make sure that it is saved using `train.report()`."
+        )
+        return []
+    else:
+        values = [result.get(key, np.nan) for result in results]
+        if not all(isinstance(value, VALID_AGGREGATE_TYPES) for value in values):
+            logger.warning(
+                f"`{key}` value type (`{type(key)}`) is not valid, "
+                f"so it is ignored. "
+                f"Make sure that its type is one of {VALID_AGGREGATE_TYPES}."
+            )
+            return []
+        return values
+
+
+class AggregateResultsPreprocessor(ResultsPreprocessor):
+    """A preprocessor that aggregates training metrics from all workers.
+
+    Args:
+        aggregation_fn (AggregateFn):
+            An aggregation method that performs the aggregation on results.
+        keys (Optional[List[str]]):
+            A list of keys reported in results to be aggregated. Keys should be saved
+            using `train.report()`.
+    """
+
+    def __init__(self, aggregation_fn: AggregateFn, keys: Optional[List[str]] = None):
+        self.aggregate_fn = aggregation_fn
+        self.keys = keys
+
+    def preprocess(self, results: List[Dict] = None) -> List[Dict]:
```

Comment:
`results` should never be `Optional` (`None`).

```suggestion
    def preprocess(self, results: List[Dict]) -> List[Dict]:
```